### PR TITLE
adding retry and error handling in charon updates

### PR DESCRIPTION
**Pull request by Galithil**
This should handle the case where the network lets us down on charon updates

### CODE REVIEW

File: ngi_pipeline/database/filesystem.py [hunk 1]
```diff
@@ -160,9 +166,18 @@ def create_charon_entries_from_project(project, best_practice_analysis="whole_ge
                                      'seqrun "{}" already exists; next...'.format(project, sample,
                                                                                   libprep, seqrun))
                     else:
+                        update_failed=True
                         LOG.error(e)
                         continue
 
+    if update_failed :
+        if retry_on_fail:
+        create_charon_entries_from_project(project, best_practice_analysis=best_practice_analysis,
```
Comment (vezzi) [inline ngi_pipeline/database/filesystem.py]: should not be indented?

File: ngi_pipeline/database/filesystem.py [hunk 2]
```diff
@@ -10,7 +10,8 @@
 
 def create_charon_entries_from_project(project, best_practice_analysis="whole_genome_reseq",
                                        sequencing_facility="NGI-S",
-                                       force_overwrite=False, delete_existing=False):
+                                       force_overwrite=False, delete_existing=False,
+                                       retry_on_fail=True):
     """Given a project object, creates the relevant entries in Charon.
     This code is remarkably shoddy as I created it in a hurry and then later
     it became a part of the pipeline. Use at your own risk! Ha ha.
```
Comment (vezzi) [inline ngi_pipeline/database/filesystem.py]: mario or denis?

File: ngi_pipeline/database/filesystem.py [hunk 3]
```diff
@@ -160,9 +166,18 @@ def create_charon_entries_from_project(project, best_practice_analysis="whole_ge
...
+                        update_failed=True
                         LOG.error(e)
                         continue
 
+    if update_failed :
+        if retry_on_fail:
+        create_charon_entries_from_project(project, best_practice_analysis=best_practice_analysis,
+                                       sequencing_facility=sequencing_facility
+                                       force_overwrite=force_overwrite, delete_existing=delete_existing,
+                                       retry_on_fail=False):
```
Comment (vezzi) [inline ngi_pipeline/database/filesystem.py]: nice... recursive function with different parameters..... it should work but it is really a trick in a pretty tricky function.
what about using something like this https://julien.danjou.info/blog/2015/python-retrying

File: ngi_pipeline/database/filesystem.py (hunk 2 above)
Comment (Galithil) [inline ngi_pipeline/database/filesystem.py]: not me. I wouldnt be that nice, you know it.

File: ngi_pipeline/database/filesystem.py (hunk 3 above)
Comment (Galithil) [inline ngi_pipeline/database/filesystem.py]: uses a non-standard package _and_ decorators. It's going to make it even harder to understand for whoever will take up on this, either rickard or szylveszter.

Comment (vezzi) [inline ngi_pipeline/database/filesystem.py]: Agree, that was the first google hit I found yesterday while travelling back from uppsala... Maybe there is something more standard. I tried to think about possible race conditions when multiple instances are called and I am not able to find any... but still I am not 100% sure this recursive call is always doing what is expected

Comment (Galithil) [inline ngi_pipeline/database/filesystem.py]: It's not real recursion, as it's impossible for it to be called more than twice. And that's on purpose.

Comment (vezzi) [inline ngi_pipeline/database/filesystem.py]: this is tail recursion.... which is  a particular case of recursion.
but ok, it is correct and should work as intended. did you checked the indentation problem? if that is fine I can merge

File: ngi_pipeline/database/filesystem.py (hunk 1 above)
Comment (Galithil) [inline ngi_pipeline/database/filesystem.py]: it's only github showing the line as a break, if you look, there's no line number.

**Pull request merged**